### PR TITLE
[Transaction Async get] Part 6: Catch unintentional switches between blocking and non-blocking calls during tests.

### PR DESCRIPTION
**Pull request by OStevan**
**Goals (and why)**:
Adds tests to prevent unintentional switching from blocking to non-blocking paths. Previously some paths were missed.

**Implementation Description (bullets)**:
- `Transaction` and `KeyValueService` are wrapped such that we keep track of the path we are on. 

**Testing (What was existing testing like?  What have you done to improve it?)**:
- these tests should make it harder for further additions to break the intended behaviour
- should give a template how to add more async paths

**Concerns (what feedback would you like?)**:
- have I covered everything
- are the changes to `SnapshotTransactionManager` allowed
- should I use something 

**Where should we start reviewing?**:
- `SnapshotTransactionTest`

**Priority (whenever / two weeks / yesterday)**:
- this week

### CODE REVIEW

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 1]
```diff
@@ -92,7 +89,24 @@ public void setUp() throws Exception {
...
-        serializableTxManager = wrapTestTransactionManager(new TestTransactionManagerImpl(
+        serializableTxManager = constructTestTransactionManager();
+
+        txManager = new CachingTestTransactionManager(serializableTxManager);
+    }
+
+    protected TrackingKeyValueService setUpKeyValueService() {
+        KeyValueService kvs = wrapKeyValueService(getBaseKeyValueService());
+        keyValueServiceWithStats = new StatsTrackingKeyValueService(kvs);
+        keyValueService = spy(new TrackingKeyValueService(keyValueServiceWithStats));
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]: this is a bad refactoring/at the very least it's confusing.

Have this method take in the `baseKeyValueService`, and then *don't* set the field in this method, and return the wrapped kvs

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 2]
```diff
@@ -92,7 +89,24 @@ public void setUp() throws Exception {
     }
 
     private void setUpTransactionManagers() {
-        serializableTxManager = wrapTestTransactionManager(new TestTransactionManagerImpl(
+        serializableTxManager = constructTestTransactionManager();
+
+        txManager = new CachingTestTransactionManager(serializableTxManager);
+    }
+
+    protected TrackingKeyValueService setUpKeyValueService() {
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]:
```suggestion
    protected KeyValueService setUpKeyValueService() {
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 3]
```diff
@@ -92,7 +89,24 @@ public void setUp() throws Exception {
...
+        KeyValueService kvs = wrapKeyValueService(getBaseKeyValueService());
+        keyValueServiceWithStats = new StatsTrackingKeyValueService(kvs);
+        keyValueService = spy(new TrackingKeyValueService(keyValueServiceWithStats));
+        return keyValueService;
+    }
+
+    protected KeyValueService wrapKeyValueService(KeyValueService trackingKeyValueService) {
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]:
```suggestion
    protected KeyValueService wrapKeyValueService(KeyValueService keyValueService) {
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java [hunk 4]
```diff
@@ -92,7 +89,24 @@ public void setUp() throws Exception {
...
+        return trackingKeyValueService;
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]:
```suggestion
        return keyValueService;
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 5]
```diff
@@ -154,14 +162,20 @@
     @Parameterized.Parameters(name = "{0}")
     public static Collection<Object[]> data() {
         Object[][] data = new Object[][] {
-                {SYNC, UnaryOperator.identity()},
-                {ASYNC, (UnaryOperator<Transaction>) GetAsyncDelegate::new}
+                {SYNC,
+                 (BiFunction<Transaction, AtomicInteger, Transaction>) (transaction, counter) -> transaction,
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Let's introduce an interface, the type parameters are getting in the way:

```java
interface WrapperWithCounter<T> {
  T apply(T toWrap, AtomicInteger counter);
}
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 6]
```diff
@@ -295,8 +313,20 @@ public void setUp() throws Exception {
     }
 
     @Override
-    protected TestTransactionManager wrapTestTransactionManager(TestTransactionManager testTransactionManager) {
-        return new WrappingTestTransactionManagerImpl(testTransactionManager, transactionWrapper);
+    protected TestTransactionManager constructTestTransactionManager() {
+        return new WrappingTestTransactionManagerImpl(
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: why do we need to have a `constructTetsTransactionManager`, is it not sufficient to wrap as before? somewhat confused

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 7]
```diff
@@ -422,28 +456,32 @@ public void testPutCleanup() throws Exception {
...
-                TransactionReadSentinelBehavior.THROW_EXCEPTION,
-                false,
-                timestampCache,
-                getRangesExecutor,
-                defaultGetRangesConcurrency,
-                MultiTableSweepQueueWriter.NO_OP,
-                MoreExecutors.newDirectExecutorService(),
-                true,
-                () -> transactionConfig));
+        AtomicInteger counter = new AtomicInteger();
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: from what I can see, you don't really do anything with the counter besides pass it in, is it not sufficient to just include it in the delegate?

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 8]
```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
...
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            transactionConfig),
+                    counter);
+        }
+    }
+
+    private static class VerifyingDelegate implements AutoDelegate_KeyValueService {
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]:
```suggestion
    private static class VerifyingKeyValueServiceDelegate implements AutoDelegate_KeyValueService {
```

Comment (felixdesouza) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: change the constructors etc.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java [hunk 9]
```diff
@@ -41,11 +51,10 @@ public Transaction delegate() {
...
         try {
-            return super.getAsync(tableRef, cells).get();
-        } catch (InterruptedException e) {
-            throw Throwables.rewrapAndThrowUncheckedException(e);
-        } catch (ExecutionException e) {
-            throw Throwables.rewrapAndThrowUncheckedException(e.getCause());
+            enterFunction.run();
+            return AtlasFutures.getUnchecked(delegate().getAsync(tableRef, cells));
+        } finally {
+            exitFunction.run();
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java]: This is possibly correct, but is a bit of a non-standard pattern: if `enterFunction.run()` throws, do we still want `exitFunction` to run? This depends on the semantics you want. Since it's tests it probably doesn't matter too much.

If we adopt Felix's suggestion (which I agree with), then this is a non-issue.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java [hunk 10]
```diff
@@ -18,19 +18,29 @@
...
+import com.palantir.atlasdb.futures.AtlasFutures;
 import com.palantir.atlasdb.keyvalue.api.Cell;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.transaction.api.Transaction;
-import com.palantir.common.base.Throwables;
 
 public class GetAsyncDelegate extends ForwardingTransaction {
 
     private final Transaction delegate;
+    private final Runnable enterFunction;
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java]: so I suppose I'm also confused, we have a counter to make sure that we're tracking synchronous gets, are we running things in parallel such that we expect this to be > 1? I don't believe `get` is at any point recursive?

[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java]: I wouldn't be opposed to even having an `AtomicInteger` that is unused in the cases where you don't care about, since this lives in tests and I wouldn't be too concerned about perf.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 11]
```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
...
+        private final KeyValueService delegate;
+        private final AtomicInteger atomicInteger;
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: This should ideally be better named, it's a bit hard to track what's going on / how this relates to things being synchronous or asynchronous (e.g. reading this class out of context would be super confusing).

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 12]
```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
...
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            () -> TRANSACTION_CONFIG),
+                    counter);
+        }
+
+        @Override
+        protected Transaction createTransaction(
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: how come there's no wrapping going on here? what's the point of the counter then even?

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 13]
```diff
@@ -295,8 +313,20 @@ public void setUp() throws Exception {
...
+                timestampService,
+                timestampService,
+                lockClient,
+                lockService,
+                transactionService,
+                conflictDetectionManager,
+                sweepStrategyManager,
+                sweepQueue,
+                MoreExecutors.newDirectExecutorService(),
+                transactionWrapper, keyValueServiceWrapper);
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: nit
```suggestion
                transactionWrapper, 
                keyValueServiceWrapper);
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 14]
```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
         return (SnapshotTransaction) transaction;
     }
 
-    private static class WrappingTestTransactionManagerImpl extends WrappingTestTransactionManager {
-        private final Function<Transaction, Transaction> transactionWrapper;
+    private static class WrappingTestTransactionManagerImpl extends TestTransactionManagerImpl {
+        private final BiFunction<Transaction, AtomicInteger, Transaction> transactionWrapper;
```
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: At some point I'm starting to think that we should just have a special `UnaryOperator<Transaction>` that is a bit more descriptive. To me it's hard to follow what the `AtomicInteger`s are trying to do. Might be worth discussing offline

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 8 above)
Comment (jeremyk-91) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]:
```suggestion
    private static class VerifyingDelegate implements ForwardingKeyValueService {
```

otherwise this could break workflows reliant on `getDelegates`

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 15]
```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
...
+                            transactionConfig),
+                    counter);
+        }
+    }
+
+    private static class VerifyingDelegate implements AutoDelegate_KeyValueService {
+        private final KeyValueService delegate;
+        private final AtomicInteger atomicInteger;
+
+        VerifyingDelegate(KeyValueService keyValueService, AtomicInteger atomicInteger) {
```
Comment (felixdesouza) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: I think we should extract a class that wraps the `AtomicInteger`, and the methods are self describing,

e.g.

```java
interface SynchronousGetTracker {
  void checkInAsyncCall();
  void checkInSyncCall();
}
```

something along those lines, having the `counter` spread over a 1000 lines is not great

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 5 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Wrapper.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 6 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Wrapping doesn't work since it when we switch to delegate we lose control how transactions are created. Refactored `TestTransactionManagerImpl` to solve this.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 14 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: This is removed and we have an interface now for this.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 12 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Mistake, fixed.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 8 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Changed.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 11 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Changed to interface and now has a better name.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 15 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: Switched to an interface.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java (hunk 9 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java]: In general we don't expect this `enterFunction.run()` to throw, now it should be clearer with the refactoring also curious about what you think about this specific part.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (OStevan) [reply atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/GetAsyncDelegate.java]: This is now resolved with an interface.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java (hunk 1 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/AtlasDbTestCase.java]: Wrapping is now also obsolete since it is done differently. Removed `keyValueServiceWithStats`.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 7 above)
Comment (OStevan) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: In general it has to be shared, even now with an interface.
